File: base_java/src/main/java/com/hesheng1024/base_java/ui/BasePresenter.java
package com.hesheng1024.base_java.ui;

import androidx.annotation.NonNull;

/**
 * @author hesheng1024
 * @date 2019/5/22 17:04
 */
public abstract class BasePresenter<V extends IBaseView, M extends IBaseModel> {
    protected final V mView;
    protected final M mModel;

    public BasePresenter(@NonNull V view) {
        this.mView = view;
        mModel = createModel();
    }

    /**
     * create model
     *
     * @return model
     */
    protected abstract M createModel();
}

File: base_kotlin/src/main/java/com/hesheng1024/base_kotlin/ui/BasePresenter.kt
package com.hesheng1024.base.base

import androidx.annotation.NonNull

/**
 *
 * @author hesheng1024
 * @date 2020/2/7 17:01
 */
abstract class BasePresenter<out V : IBaseView, out M : IBaseModel>(@NonNull protected val mView: V) {
    protected val mModel = this.createModel()

    abstract fun createModel(): M
}